File: java/src/test/java/YatzyTest.java
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class YatzyTest {

    @Test
    public void chance_scores_sum_of_all_dice() {
        assertThat(new Yatzy(2, 3, 4, 5, 1).chance()).isEqualTo(15);
        assertThat(new Yatzy(3, 3, 4, 5, 1).chance()).isEqualTo(16);
    }

    @Test
    public void yatzy_scores_50() {
        assertThat(new Yatzy(4, 4, 4, 4, 4).yatzy()).isEqualTo(50);
        assertThat(new Yatzy(6, 6, 6, 6, 6).yatzy()).isEqualTo(50);
        assertThat(new Yatzy(6, 6, 6, 6, 3).yatzy()).isEqualTo(0);
    }

    @Test
    public void test_1s() {
        assertThat(new Yatzy(1, 2, 3, 4, 5).ones()).isEqualTo(1);
        assertThat(new Yatzy(1, 2, 1, 4, 5).ones()).isEqualTo(2);
        assertThat(new Yatzy(6, 2, 2, 4, 5).ones()).isEqualTo(0);
        assertThat(new Yatzy(1, 2, 1, 1, 1).ones()).isEqualTo(4);
    }

    @Test
    public void test_2s() {
        assertThat(new Yatzy(1, 2, 3, 2, 6).twos()).isEqualTo(4);
        assertThat(new Yatzy(2, 2, 2, 2, 2).twos()).isEqualTo(10);
    }

    @Test
    public void test_threes() {
        assertThat(new Yatzy(1, 2, 3, 2, 3).threes()).isEqualTo(6);
        assertThat(new Yatzy(2, 3, 3, 3, 3).threes()).isEqualTo(12);
    }

    @Test
    public void fours_test() {
        assertThat(new Yatzy(4, 4, 4, 5, 5).fours()).isEqualTo(12);
        assertThat(new Yatzy(4, 4, 5, 5, 5).fours()).isEqualTo(8);
        assertThat(new Yatzy(4, 5, 5, 5, 5).fours()).isEqualTo(4);
    }

    @Test
    public void fives() {
        assertThat(new Yatzy(4, 4, 4, 5, 5).fives()).isEqualTo(10);
        assertThat(new Yatzy(4, 4, 5, 5, 5).fives()).isEqualTo(15);
        assertThat(new Yatzy(4, 5, 5, 5, 5).fives()).isEqualTo(20);
    }

    @Test
    public void sixes_test() {
        assertThat(new Yatzy(4, 4, 4, 5, 5).sixes()).isEqualTo(0);
        assertThat(new Yatzy(4, 4, 6, 5, 5).sixes()).isEqualTo(6);
        assertThat(new Yatzy(6, 5, 6, 6, 5).sixes()).isEqualTo(18);
    }

    @Test
    public void one_pair() {
        assertThat(new Yatzy(3, 4, 3, 5, 6).pair()).isEqualTo(6);
        assertThat(new Yatzy(5, 3, 3, 3, 5).pair()).isEqualTo(10);
        assertThat(new Yatzy(5, 3, 6, 6, 5).pair()).isEqualTo(12);
    }

    @Test
    public void two_pairs() {
        assertThat(new Yatzy(3, 3, 5, 4, 5).twoPairs()).isEqualTo(16);
        assertThat(new Yatzy(3, 3, 5, 5, 5).twoPairs()).isEqualTo(16);
    }

    @Test
    public void three_of_a_kind() {
        assertThat(new Yatzy(3, 3, 3, 4, 5).threeOfAKind()).isEqualTo(9);
        assertThat(new Yatzy(5, 3, 5, 4, 5).threeOfAKind()).isEqualTo(15);
        assertThat(new Yatzy(3, 3, 3, 3, 5).threeOfAKind()).isEqualTo(9);
    }

    @Test
    public void four_of_a_kind() {
        assertThat(new Yatzy(3, 3, 3, 3, 5).fourOfAKind()).isEqualTo(12);
        assertThat(new Yatzy(5, 5, 5, 4, 5).fourOfAKind()).isEqualTo(20);
        assertThat(new Yatzy(3, 3, 3, 3, 3).fourOfAKind()).isEqualTo(12);

    }

    @Test
    public void small_straight() {
        assertThat(new Yatzy(1, 2, 3, 4, 5).smallStraight()).isEqualTo(15);
        assertThat(new Yatzy(2, 3, 4, 5, 1).smallStraight()).isEqualTo(15);
        assertThat(new Yatzy(1, 2, 2, 4, 5).smallStraight()).isEqualTo(0);
    }

    @Test
    public void large_straight() {
        assertThat(new Yatzy(6, 2, 3, 4, 5).largeStraight()).isEqualTo(20);
        assertThat(new Yatzy(2, 3, 4, 5, 6).largeStraight()).isEqualTo(20);
        assertThat(new Yatzy(1, 2, 2, 4, 5).largeStraight()).isEqualTo(0);
    }

    @Test
    public void full_house() {
        assertThat(new Yatzy(6, 2, 2, 2, 6).fullHouse()).isEqualTo(18);
        assertThat(new Yatzy(2, 3, 4, 5, 6).fullHouse()).isEqualTo(0);
    }

}

File: javascript/test/yatzy-test.js
/* eslint-disable prefer-arrow-callback */
/* eslint-disable func-names */

const assert = require('assert');
const Yatzy = require('../lib/yatzy');


describe('Chance', function () {
    it('scores sum of all dice', function () {
        assert.equal(15, new Yatzy().chance(2, 3, 4, 5, 1));
        assert.equal(16, new Yatzy().chance(3, 3, 4, 5, 1));
    });
});

describe('Yatzy', function () {
    it('scores 50', function () {
        assert.equal(50, new Yatzy().yatzy(4, 4, 4, 4, 4));
        assert.equal(50, new Yatzy().yatzy(6, 6, 6, 6, 6));
        assert.equal(0, new Yatzy().yatzy(6, 6, 6, 6, 3));
    });
});

describe('Ones', function () {
    it('score the sum of 1s', function () {
        assert.equal(1, new Yatzy().ones(1, 2, 3, 4, 5));
        assert.equal(2, new Yatzy().ones(1, 2, 1, 4, 5));
        assert.equal(0, new Yatzy().ones(6, 2, 2, 4, 5));
        assert.equal(4, new Yatzy().ones(1, 2, 1, 1, 1));
    });
});

describe('Twos', function () {
    it('score the sum of 2s', function () {
        assert.equal(4, new Yatzy().twos(1, 2, 3, 2, 6));
        assert.equal(10, new Yatzy().twos(2, 2, 2, 2, 2));
    });
});

describe('Threes', function () {
    it('score the sum of 3s', function () {
        assert.equal(6, new Yatzy().threes(1, 2, 3, 2, 3));
        assert.equal(12, new Yatzy().threes(2, 3, 3, 3, 3));
    });
});

describe('Fours', function () {
    it('score the sum of 4s', function () {
        assert.equal(12, new Yatzy(4, 4, 4, 5, 5).fours());
        assert.equal(8, new Yatzy(4, 4, 5, 5, 5).fours());
        assert.equal(4, new Yatzy(4, 5, 5, 5, 5).fours());
    });
});

describe('Fives', function () {
    it('score the sum of fives', function () {
        assert.equal(10, new Yatzy(4, 4, 4, 5, 5).fives());
        assert.equal(15, new Yatzy(4, 4, 5, 5, 5).fives());
        assert.equal(20, new Yatzy(4, 5, 5, 5, 5).fives());
    });
});

describe('Sixes', function () {
    it('score the sum of sixes', function () {
        assert.equal(0, new Yatzy(4, 4, 4, 5, 5).sixes());
        assert.equal(6, new Yatzy(4, 4, 6, 5, 5).sixes());
        assert.equal(18, new Yatzy(6, 5, 6, 6, 5).sixes());
    });
});

describe('One pair', function () {
    it('scores the sum of the highest pair', function () {
        assert.equal(6, new Yatzy().score_pair(3, 4, 3, 5, 6));
        assert.equal(10, new Yatzy().score_pair(5, 3, 3, 3, 5));
        assert.equal(12, new Yatzy().score_pair(5, 3, 6, 6, 5));
    });
});

describe('Two pair', function () {
    it('scores the sum of the two pairs', function () {
        assert.equal(16, new Yatzy().two_pair(3, 3, 5, 4, 5));
        assert.equal(16, new Yatzy().two_pair(3, 3, 5, 5, 5));
    });
});

describe('Three of a kind', function () {
    it('scores the sum of the three of the kind', function () {
        assert.equal(9, new Yatzy().three_of_a_kind(3, 3, 3, 4, 5));
        assert.equal(15, new Yatzy().three_of_a_kind(5, 3, 5, 4, 5));
        assert.equal(9, new Yatzy().three_of_a_kind(3, 3, 3, 3, 5));
    });
});

describe('Four of a kind', function () {
    it('scores the sum of the four of the kind', function () {
        assert.equal(12, new Yatzy().four_of_a_kind(3, 3, 3, 3, 5));
        assert.equal(20, new Yatzy().four_of_a_kind(5, 5, 5, 4, 5));
        assert.equal(9, new Yatzy().three_of_a_kind(3, 3, 3, 3, 3));
    });
});

describe('Small straight', function () {
    it('scores 15', function () {
        assert.equal(15, new Yatzy().smallStraight(1, 2, 3, 4, 5));
        assert.equal(15, new Yatzy().smallStraight(2, 3, 4, 5, 1));
        assert.equal(0, new Yatzy().smallStraight(1, 2, 2, 4, 5));
    });
});

describe('Large straight', function () {
    it('scores 20', function () {
        assert.equal(20, new Yatzy().largeStraight(6, 2, 3, 4, 5));
        assert.equal(20, new Yatzy().largeStraight(2, 3, 4, 5, 6));
        assert.equal(0, new Yatzy().largeStraight(1, 2, 2, 4, 5));
    });
});

describe('Full house', function () {
    it('scores the sum of the full house', function () {
        assert.equal(18, new Yatzy().fullHouse(6, 2, 2, 2, 6));
        assert.equal(0, new Yatzy().fullHouse(2, 3, 4, 5, 6));
    });
});
